test/integration/certificates/readiness_controller_test.go

```diff
@@ -0,0 +1,185 @@
+// readiness controller sets the correct Ready condition on a Certificate.
+func TestReadinessController(t *testing.T) {
+	config, stopFn := framework.RunControlPlane(t)
+	defer stopFn()
+
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Second*20)
+	defer cancel()
+
+	// Fix the time to be able to test expired certificate scenario.
+	fakeClock := &fakeclock.FakeClock{}
+	// Build, instantiate and run the trigger controller.
+	kubeClient, factory, cmCl, cmFactory := framework.NewClients(t, config)
+
+	namespace := "testns"
+	certName := "test"
+	secretName := "test"
+	// notBefore value on the test certificate
+	notBefore := fakeClock.Now().UTC().Add(time.Hour).Truncate(time.Second)
+	// notAfter value on the test certificate
+	notAfter := notBefore.Add(time.Hour * 4).Truncate(time.Second)
+	// expected renewal time
+	renewalTime := notBefore.Add(time.Hour * 2).Truncate(time.Second)
+	issuer := cmmeta.ObjectReference{
+		Name:  "testissuer",
+		Kind:  "IssuerKind",
+		Group: "group.example.com",
+	}
+
+	// Create Namespace.
+	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}}
+	_, err := kubeClient.CoreV1().Namespaces().Create(context.TODO(), ns, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ctrl, queue, mustSync := readiness.NewController(logf.Log, cmCl, factory, cmFactory, readiness.NewReadinessPolicyChain(fakeClock), certificates.RenewalTimeWrapper(cmapi.DefaultRenewBefore), readiness.PolicyEvaluator)
+	c := controllerpkg.NewController(
+		context.Background(),
+		"readiness_test",
+		metrics.New(logf.Log),
+		ctrl.ProcessItem,
+		mustSync,
+		nil,
+		queue,
+	)
+	stopController := framework.StartInformersAndController(t, factory, cmFactory, c)
+	defer stopController()
+
+	// 1. A newly created Certificate without a Secret should not be set to Ready.
+
+	// create the Certificate
+	cert, err := cmCl.CertmanagerV1().Certificates(namespace).Create(ctx,
+		gen.Certificate(certName,
+			gen.SetCertificateNamespace(namespace),
+			gen.SetCertificateSecretName(secretName),
+			gen.SetCertificateCommonName("example.com"),
+			gen.SetCertificateIssuer(issuer),
+			gen.SetCertificateRenewBefore(time.Hour*2)),
+		metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	// ensure that Certificate does not have Ready conditon set to True
+	ensureConditionNotApplied(ctx, t, cmCl, cert, cmapi.CertificateCondition{
+		Type:   cmapi.CertificateConditionReady,
+		Status: cmmeta.ConditionTrue,
+	})
+
+	// 2. A Certificate with a valid Secret and CertificateRequest should have Ready condition set to True
+	// and updated status.RenewalTime, status.NotBefore, status.NotAfter fields
+
+	// Create private key
+	privKeyBytes := mustCreatePEMPrivateKey(t)
+	// Create x509 certificate
+	x509Cert := mustCreateCertWithNotBeforeAfter(t, privKeyBytes, cert, notBefore, notAfter)
+	// Create a Secret
+	secret, err := kubeClient.CoreV1().Secrets(namespace).Create(ctx,
```

Comment:
suggestion: skip the "paraphrasing" comment 😅

```suggestion

	privKeyBytes := mustCreatePEMPrivateKey(t)
	x509Cert := mustCreateCertWithNotBeforeAfter(t, privKeyBytes, cert, notBefore, notAfter)
	secret, err := kubeClient.CoreV1().Secrets(namespace).Create(ctx,
```